src/strands/experimental/tools/mcp/mcp_tool_provider.py

```diff
@@ -0,0 +1,171 @@
+"""MCP Tool Provider implementation."""
+
+import logging
+from typing import Any, Callable, Optional, Pattern, Sequence, Union
+
+from typing_extensions import TypedDict
+
+from ....tools.mcp.mcp_agent_tool import MCPAgentTool
+from ....tools.mcp.mcp_client import MCPClient
+from ....types.exceptions import ToolProviderException
+from ....types.tools import AgentTool
+from ..tool_provider import ToolProvider
+
+logger = logging.getLogger(__name__)
+
+_ToolFilterCallback = Callable[[AgentTool], bool]
+_ToolFilterPattern = Union[str, Pattern[str], _ToolFilterCallback]
```

Comment:
nit - use the new pipe syntax for unions

```suggestion
_ToolFilterCallback = Callable[[AgentTool], bool]
_ToolFilterPattern = str | Pattern[str] | _ToolFilterCallback
```